### DISCUSSION TITLE
Stream/Observable

### DISCUSSION THREAD

**Discussion by darky** [mikro-orm/mikro-orm#4436] [General]
**Describe the solution you'd like**
Would be nice to have some methods, which return stream/observable

**Describe alternatives you've considered**
* https://www.npmjs.com/package/pg-query-stream
* https://knexjs.org/guide/interfaces.html#streams
* https://typeorm.delightful.studio/classes/_driver_postgres_postgresqueryrunner_.postgresqueryrunner.html#stream

**Comment by B4nan** [comment]
I am not sure if that would be technically possible, unlike all the libraries you mentioned, MikroORM maintains an in-memory state, you can't stream that.

Do you have an actual use case, or it would be just _nice to have_? :]

**Reply by darky** [reply]
Mostly *nice to have*
Stream based applications used 2x-3x less RAM consumption (My personal bench of **pg-query-stream**)
I created this issue with hope, that it is not difficult task. While fetching need create pipe and mutate MikroORM in-memory state on the fly.
But if it's hard to implement in MikroORM directly, afraid that issue unworthy for implementing :)
Maybe I can implement it as third-party plugin? Is MikroORM expose API for managing internal state directly?

**Reply by B4nan** [reply]
Well, I have no idea how it should work, nor about the use cases, that's my biggest problem. I am open to concrete suggestions or a PR.

**Reply by TechnotronicOz** [reply]
Streaming is foundational to so many implementations, almost so much that nearly all other libraries treat it ubiquitously. At this point, it almost feels as if Sequelize is dying given the lack of traction over such a long period for a feature considered to be standard.

* http://c2fo.github.io/patio/patio_Dataset.html#stream
* https://typeorm.io/select-query-builder#streaming-result-data
* https://kysely-org.github.io/kysely-apidoc/interfaces/SelectQueryBuilder.html#stream
* https://knexjs.org/guide/interfaces.html#streams
* https://node-postgres.com/apis/cursor via pg-cursor
* https://www.npmjs.com/package/pg-query-stream

**Comment by bangbang93** [comment]
I'm now using with
```typescript
createQueryBuilder()
      .where(query)
      .orderBy({id: 1})
      .getKnexQuery()
      .stream()
```
It can be used with knex stream function. But cannot get query log.
Stream is very important for query large data set (~1M rows in a single query)

**Comment by ruscon** [comment]
It would be cool to have something like this out of the box

I have method like this in a service class (could be done in a custom repository class)

```ts
// SqlLatestDocument - my entity
// this.getRepository() - method to get entity repository in my case

import { type Cursor, type Loaded, type FilterQuery, type FindByCursorOptions } from '@mikro-orm/core'

class MyService {
  protected getRepository(): EntityRepository<SqlLatestDocument> {
     // 
  }

  async *batchStream<Hint extends string = never, Fields extends string = '*', Excludes extends string = never, I extends boolean = true>(
    where: FilterQuery<T>,
    options: FindByCursorOptions<T, Hint, Fields, Excludes, I>
  ): AsyncGenerator<Loaded<T, Hint, Fields, Excludes>[]> {
    let cursor: Cursor<T, Hint, Fields, Excludes> | undefined = undefined

    do {
      cursor = await this.findByCursor<Hint, Fields, Excludes>(where, {
        // @ts-expect-error type issue with mikro-orm
        includeCount: false,
        first: 100,
        ...options,
        after: cursor
      })

      yield cursor.items
    } while (cursor.hasNextPage)
  }

  async *stream<Hint extends string = never, Fields extends string = '*', Excludes extends string = never, I extends boolean = true>(
    where: FilterQuery<T>,
    options: FindByCursorOptions<T, Hint, Fields, Excludes, I>
  ): AsyncGenerator<Loaded<T, Hint, Fields, Excludes>> {
    for await (const items of this.batchStream<Hint, Fields, Excludes, I>(where, options)) {
      for (const item of items) {
        yield item
      }
    }
  }
}
```

How to use
```ts
const service = new MyService()

for await (const document of service.stream(where, options)) {
 //
}
```

**Reply by B4nan** [reply]
>         // @ts-expect-error type issue with mikro-orm

sidenote: this is issue in your code, you are not using the generic `I` type in the `findByCursor` call

```ts
cursor = await this.getRepository().findByCursor<Hint, Fields, Excludes, I>(where, {
```

**Reply by ruscon** [reply]
@B4nan There is no such possibility in the `EntityRepository`
https://github.com/mikro-orm/mikro-orm/blob/4ab159de1d91cc9a7670221e1cfc148722fc2ddf/packages/core/src/entity/EntityRepository.ts#L152-L156

Only in the EntityManager https://github.com/mikro-orm/mikro-orm/blob/4ab159de1d91cc9a7670221e1cfc148722fc2ddf/packages/core/src/EntityManager.ts#L657-L663

**Reply by B4nan** [reply]
Oh I see. PR welcome for that.

**Reply by B4nan** [reply]
34c963086e37c059ec9d2ece1fe988bfbeca2967

**Comment by ruscon** [comment]
What about `steam`? Do you have better solution or mine is good ?

**Reply by B4nan** [reply]
Not really, I always considered the streaming support as something more low-level (e.g. streaming rows from the database on knex/kysely level), but then I wasn't sure how it could work, we can't just process a single line, not outside of raw results.

**Reply by ruscon** [reply]
Well, you have the .map method, which converts to an entity. Can't you do it properly with it?

In normal projects where there are more than 100 records in a table, it is dangerous to work without real stream/cursor.
I don't see any point in switching to native methods, otherwise why would we need orm?

**Reply by B4nan** [reply]
It wouldn't work for to-many relations, those can yield multiple rows that are merged into a single entity graph, and you have no control over how many rows you need for a single root entity.

**Comment by B4nan** [comment]
First draft for native streaming support in v7: #6923

**Comment by ruscon** [comment]
Thank you, will check

**Reply by B4nan** [reply]
Note that this won't work for cursor-based pagination, since we need to return the results in a reversed order sometimes, and we can't do that when we stream.

https://github.com/mikro-orm/mikro-orm/blob/master/packages/knex/src/AbstractSqlDriver.ts#L152

Maybe we could do something like what you highlighted above with streaming the batches for this use case. Btw one note to your implementation - unless you clear the context somewhere, this will still keep everything in the identity map, which will grow over time and end up with OOM eventually. You don't want a persistent context for streaming, `findByCursor` should be called with `disableIdentityMap: true`.
